Newton merge 3: To the spd factor visualisation and (a bit too far) beyond

MFraters opened the pull request:
This pull request was intended to only add a visualisation post processor for the spd factor computations. But to make that work I also needed material models to play with. 

So I have rewritten drucker prager (@anne-glerum, are you oke with this?) to be able to compute the derivatives (analytically) and have written tests for them. The tests test the derivative of the material model (which are analytical) with a finite difference derivative. There are some cases where there are differences between these, which I can derive from both sides, but don't know which is actually better.

gassmoeller commented:
/run-tests

bangerth commented:
Can you fix the tests?

bangerth commented on include/aspect/material_model/drucker_prager.h:
```diff
@@ -169,14 +144,14 @@ namespace aspect
          * The angle of internal friction
          */
-        double phi;
+        double angle_of_internal_friction;
```
nice job using descriptive names

bangerth commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,108 @@
+    /**
+     * A material model based on a simple powerlaw rheology and
+     * Implements the derivatives needed for the Newton method.
+     * TODO: more documentation.
```
can you add this? In particular, can you add a formula that describes the viscosity?

bangerth commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,108 @@
+         */
+        std::vector<double> min_strain_rate;
+        std::vector<double> min_visc;
+        std::vector<double> max_visc;
```
spell it out: `min/max_viscosity`.

MFraters commented:
working on it, but it has to do with that the rewriting of the drucker prager material model uses slightly different assumptions.

bangerth commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,108 @@
+
+        unsigned int n_fields;
+        // averaging parameters
+        double viscosity_averaging_p;
```
these are a lot of parameters. It will be that much more important to have a formula (in the class doc, and in particular in the documentation at the bottom of the .cc file) that explains how they all fit together.

You may then go back to the documentation of all of these member variables and explain what symbols in the formula they correspond to.

bangerth commented on include/aspect/material_model/simple_nonlinear.h:
```diff
@@ -0,0 +1,108 @@
+        double viscosity_averaging_p;
+
+
+    };
```
no need for the empty lines

bangerth commented on include/aspect/postprocess/visualization/spd_factor.h:
```diff
@@ -0,0 +1,62 @@
+       * base class. See there for their meaning.
+       */
+      template <int dim>
+      class SPDfactor
```
Write it `SPDFactor`, or `SPD_Factor`. "factor" is a new word, it should be capitalized in class names.

bangerth commented on source/material_model/drucker_prager.cc:
```diff
@@ -29,56 +31,116 @@ namespace aspect
+
+                  derivatives->viscosity_derivative_wrt_pressure[i] = composition_viscosity_pressure_derivatives;//Utilities::derivatives_weighed_p_norm_average(out.viscosities[i],volume_fractions, composition_viscosities, composition_viscosities_pressure_derivatives, viscosity_averaging_p);
+
+#ifdef DEBUG
```
no need for this -- in release mode, the `Assert` statements will be replaced by empty statements (=nothing), and then the compiler will recognize that the double loop does nothing and remove it as well.

bangerth commented on source/material_model/drucker_prager.cc:
```diff
@@ -29,56 +31,116 @@ namespace aspect
+        }
     }
 
-
```
we like 3 empty lines between functions

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
```
2017

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+  namespace
+  {
+    std::vector<double>
+    get_vector_double (const std::string &parameter, const unsigned int n_fields, ParameterHandler &prm)
```
we don't use this function any more. there is a replacement that came in via #1030 (see also #528).

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+    {
+      /*std::cout << "vfFlag 1 " << std::endl;
+         std::vector<double> volume_fractions(compositional_fields.size());
+         std::cout << "vfFlag 2 " << std::endl;
```
if this debug output is still helpful, it would be worthwhile writing it in the style of properly indented code. But I suspect that you can get rid of it.

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+          volume_fractions[0] = 1.0 - sum_composition; //background mantle
+          for ( unsigned int i=1; i <= x_comp.size(); ++i)
+            volume_fractions[i] = x_comp[i-1];
+        } //std::cout << "Background mantle = " << volume_fractions[0] << std::endl;
```
remove debug output

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+          // const Point<dim> position = in.position[i];
```
remove the statement

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+
+          // Averaging composition-field dependent properties
+          // Compositions
+          //This assert may be help when designing tests for this material model
```
we usually leave a space after `//`

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+                        {
+                          //strictly speaking the derivative is this: 0.5 * ((1/stress_exponent)-1) * std::pow(2,2) * out.viscosities[i] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i])
+                          composition_viscosities_derivatives[c] = 2 * (stress_exponent_inv-1) * composition_viscosities[c] * (1/(edot_ii*edot_ii)) * deviator(in.strain_rate[i]);
+                          //std::cout << composition_viscosities_derivatives[i] << ", dev strt:" << deviator(in.strain_rate[i]) << std::endl;
```
remove

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+                      else
+                        {
+                          composition_viscosities_derivatives[c] = 0;
+                          //std::cout << "Set " << i << " to 0" << std::endl;
```
remove

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+              Assert(dealii::numbers::is_finite(out.viscosities[i]),ExcMessage ("Error: Averaged viscosity is not finite."));
+
+              //for(int ccc = 0; ccc < in.composition[i].size(); ccc++)
+              //  std::cout << " comp " << ccc << " = " << in.composition[i][ccc] << std::endl;
```
remove

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+                  for (int x = 0; x < dim; x++)
+                    for (int y = 0; y < dim; y++)
+                      if (!dealii::numbers::is_finite(derivatives->viscosity_derivative_wrt_strain_rate[i][x][y]))
+                        std::cout << "Error: Averaged viscosity to strain-rate devrivative is not finite." << std::endl;
```
convert this into an `Assert` -- nothing good can happen if the condition is not true.

same below. You can then remove the `#ifdef DEBUG` as well

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+    SimpleNonlinear<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      // can't use this->n_compositional_fields(), because some
+      // tests never initiate the simulator, but uses the material
```
-> initialize

bangerth commented on source/material_model/simple_nonlinear.cc:
```diff
@@ -0,0 +1,435 @@
+      prm.enter_subsection ("Compositional fields");
+      {
+        n_fields = prm.get_integer ("Number of fields")+1;
+        //AssertThrow(n_fields > 0, ExcMessage("This material model needs at least one compositional field."))
```
if you don't need it, remove it. otherwise leave it in but without the comment marker

bangerth commented on source/postprocess/visualization/spd_factor.cc:
```diff
@@ -0,0 +1,113 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
2017

MFraters commented:
I have addressed nearly all comments, will do the rest tomorrow. One quick comment on the tests: I have adjusted everything tot the way it used to be done in this material model except for one thing, the definition of the second invariant. In theory it shouldn't matter for incompressible cases (if you use it together with fabs and sqrt). I think that it is incorrect to use the deal.ii function second invariant in any geodynamics material model, so I use the sqrt(0.5*strain_rate*strain_rate). We can discuss this tomorrow.

bangerth commented:
My suggestion would be to do this in a separate patch -- write the new functionality now, let us get that merged (hopefully with no changes to the output files), and then make the change in the definition of the invariant in a separate patch that is small, self-contained, and leads to different results. That makes it easier to review things.

bangerth commented:
It doesn't compile right now.

MFraters commented:
This pull request has been split up into the pull requests #1683, #1684 and #1686. This pull request can therefore be closed.
